datalad/__init__.py

```diff
@@ -19,6 +19,68 @@
     raise RuntimeError('DataLad cannot run in "optimized" mode, i.e. python -O')
 
 
+# this is not to be modified. for querying use get_apimode()
+__api = 'python'
+# this is not to be modified. see enable/in_librarymode()
+# by default, we are in application-mode, simply because most of
+# datalad was originally implemented with this scenario assumption
+__runtime_mode = 'application'
+
+
+def get_apimode():
+    """Returns the API mode label for the current session.
+
+    The API mode label indicates whether DataLad is running in "normal"
+    mode in a Python session, or whether it is used via the command line
+    interface.
+
+    This function is a utility for optimizing behavior and messaging to the
+    particular API (Python vs command line) in use in a given process.
+
+    Returns
+    {'python', 'cmdline'}
+      The API mode is 'python' by default, unless the main command line
+      entrypoint set it to 'cmdline'.
+    """
+    return __api
+
+
+def in_librarymode():
+    """Returns whether DataLad is requested to run in "library mode"
+
+    In this mode DataLad aims to behave without the assumption that it is
+    itself the front-end of a process and in full control over messaging
+    and parameters.
+
+    Returns
+    -------
+    bool
+    """
+    return __runtime_mode == 'library'
+
+
+def enable_librarymode():
+    """Request DataLad to operate in library mode.
+
+    This function should be executed immediately after importing the `datalad`
+    package, when DataLad is not used as an application, or in interactive
+    scenarios, but as a utility library inside other applications. Enabling
+    this mode will turn off some convenience feature that are irrelevant in
+    such use cases (with performance benefits), and alters it messaging
+    behavior to better interoperate with 3rd-party front-ends.
+
+    Library mode can only be enabled once. Switching it on and off within
+    the runtime of a process is not supported.
+
+    Example::
+
+        >>> import datalad
+        >>> datalad.enable_librarymode()
+    """
+    global __runtime_mode
+    __runtime_mode = 'library'
```

Comment:
I worry a little of placing "library" etc into function names .
I wonder if it wouldn't be better to encapsulate it in a helper class (e.g. `_DataLadMode`) with `.current`, `.supported` (returns available -- we might need to add later some other, e.g. `jupyter-shell`, `web-app` although since it relates to `ui`s - not yet sure if separation would be clear since that is where we also we might need to adjust UI - e.g. use jupyter widgets for progress bars), `.switch_to(mode: str)` and have an instance `datalad.mode` exposing it? Could also have `__eq__` overloaded to implement `self.current == other`.

Reply:
I am not confident I get your argument properly.

> I worry a little of placing "library" etc into function names

I am OK with a different name, which name would you suggest?

> I wonder if it wouldn't be better to encapsulate it in a helper class

I proposed three simple functions to be able to have maximum freedom for such changes, once they turn out to be necessary. They can *internally* do what you proposed.

However, so far the concept is only developed to the degree that we know we need to distinguish an API mode (with a presently fixed set of two possibilities, without preventing future extension) and standalone vs library-mode (whatever labels we end up using for them). We have not yet talked about any mode switching, additional modes, etc. If we need to add them having a `in_librarymode()` helper is not an obstacle.

All this would have implications on how to test for such conditions throughout the code-base (and other people's code). I feel confident to be able to support these three functions indefinitely. It is not clear to me whether you have maintenance concerns. From my reading I'd say that everything you write can be done in the future on top of these changes. But if I am missing something here, we likely need to close this PR again, and work on a better concept. I only started implementing this, after going through the issues and finding that we did not come up with additional needs for some years.